File: payment-matcher/src/utils/amountGenerator.ts
/**
 * Unique payment amount generator.
 *
 * Formula: uniqueAmount = baseAmount + tip, tip ∈ [1, 99]
 * Guarantees the result is not in the `taken` set (active pending intents).
 *
 * Used by the FastAPI backend in production; kept here for parity / tests /
 * optional standalone tooling.
 */
export function generateUniqueAmount(baseAmount: number, taken: Set<number>): number {
  if (!Number.isInteger(baseAmount) || baseAmount < 1) {
    throw new Error("baseAmount must be a positive integer (so'm)");
  }

  const tips = Array.from({ length: 99 }, (_, i) => i + 1);
  // Fisher–Yates shuffle
  for (let i = tips.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [tips[i], tips[j]] = [tips[j], tips[i]];
  }

  for (const tip of tips) {
    const amount = baseAmount + tip;
    if (!taken.has(amount)) return amount;
  }

  throw new Error("No free unique amount available (all tips 1–99 taken)");
}

File: gobron-backend/app/utils/amount_generator.py
"""Unique payment amount = base + tip (1..99).

Two concurrent pending intents must never share the same unique_amount.
"""
from __future__ import annotations

import random


def generate_unique_amount(base_amount: int, taken: set[int]) -> int:
    """Return base_amount + tip in [1, 99] not present in ``taken``.

    Raises ValueError if every tip is already taken (extremely rare).
    """
    if base_amount < 1:
        raise ValueError("base_amount must be positive")

    candidates = list(range(1, 100))
    random.shuffle(candidates)
    for tip in candidates:
        amount = base_amount + tip
        if amount not in taken:
            return amount
    raise ValueError("No free unique amount available; try again shortly")
